EIPS/eip_generic_token_upgrade_standard.md

```diff
@@ -0,0 +1,353 @@
+    /// The call will fail if upgrade status is not true, if approve has not been called
+    /// on the source contract, or if sourceAmount is larger than the amount of source tokens at the msg.sender address.
+    /// If the ratio would cause an amount of tokens to be destroyed by rounding/truncation, the upgrade call will
+    /// only upgrade the nearest whole amount of source tokens returning the excess to the msg.sender address. 
+    /// Emits the Upgrade event
+    /// @param _to The address the destination tokens will be sent to upon completion of the upgrade
+    /// @param sourceAmount The amount of source tokens that will be upgraded 
+    function upgrade(address _to, uint256 sourceAmount) external {
+	require(upgradeStatus == true, "SourceUpgrade: upgrade status is not active");
+	(uint256 destinationAmount, uint256 sourceRemainder) = computeUpgrade(sourceAmount);
+	sourceAmount -= sourceRemainder;
+	require(sourceAmount > 0, "SourceUpgrade: disallow conversions of zero value");
+	
+	upgradedBalance[msg.sender] += sourceAmount;
+	source.safeTransferFrom(
+				msg.sender,
+				address(this),
+				sourceAmount
+				);
+	destination.safeTransfer(_to, destinationAmount);
+	sourceUpgradedTotal += sourceAmount;
+	emit Upgrade(msg.sender, _to, sourceAmount, destinationAmount);
+    }
+    
+    /// @dev A method to conduct a downgrade from destination token to source token.
+    /// The call will fail if downgrade status is not true, if approve has not been called
+    /// on the destination contract, or if destinationAmount is larger than the amount of destination tokens at the msg.sender address.
+    /// If the ratio would cause an amount of tokens to be destroyed by rounding/truncation, the downgrade call will only downgrade
+    /// the nearest whole amount of destination tokens returning the excess to the msg.sender address. 
+    ///  Emits the Downgrade event
+    /// @param _to The address the source tokens will be sent to upon completion of the downgrade
+    /// @param destinationAmount The amount of destination tokens that will be downgraded 
+    function downgrade(address _to, uint256 destinationAmount) external {
+	require(upgradeStatus == true, "SourceUpgrade: upgrade status is not active");
+	(uint256 sourceAmount, uint256 destinationRemainder) = computeDowngrade(destinationAmount);
+	destinationAmount -= destinationRemainder;
+	require(destinationAmount > 0, "SourceUpgrade: disallow conversions of zero value");
+	require(upgradedBalance[msg.sender] >= sourceAmount,
+		"SourceUpgrade: can not downgrade more than previously upgraded"
+		);
+	
+	upgradedBalance[msg.sender] -= sourceAmount;
+	destination.safeTransferFrom(
+				     msg.sender,
+				     address(this),
+				     destinationAmount
+				     );
+	source.safeTransfer(_to, sourceAmount);
+	sourceUpgradedTotal -= sourceAmount;
+	emit Downgrade(msg.sender, _to, sourceAmount, destinationAmount);
+    }
+}
```

Comment:
This code block has irregularly mixed tabs and spaces, please have a consistent indentation strategy.